h/views/admin/documents.py

```diff
@@ -0,0 +1,53 @@
+import json
+
+from pyramid.view import view_config, view_defaults
+
+from h.schemas import ValidationError
+from h.schemas.annotation import URLMigrationSchema
+from h.security import Permission
+from h.tasks.url_migration import move_annotations_by_url
+
+
+@view_defaults(route_name="admin.documents", permission=Permission.AdminPage.HIGH_RISK)
+class DocumentsAdminViews:
+    def __init__(self, request):
+        self.request = request
+
+    @view_config(
+        request_method="GET", renderer="h:templates/admin/documents.html.jinja2"
+    )
+    def get(self):
+        return {}
+
+    @view_config(
+        request_method="POST",
+        request_param="update_annotation_urls",
+        require_csrf=True,
+        renderer="h:templates/admin/documents.html.jinja2",
+    )
+    def update_annotation_urls(self):
+        try:
+            url_mappings_json = json.loads(self.request.params["url_mappings"])
+        except ValueError as err:
+            return self.report_error(f"Failed to parse URL mappings: {err}")
+
+        schema = URLMigrationSchema()
+
+        try:
+            url_mappings = schema.validate(url_mappings_json)
+        except ValidationError as err:
+            return self.report_error(f"Failed to validate URL mappings: {err}")
+
+        # Chunk up the URLs into batches to limit the work per task.
+        move_annotations_by_url.chunks(url_mappings.items(), 10).apply_async()
+
+        self.request.session.flash(
+            f"URL migration started for {len(url_mappings)} URL(s)", "success"
+        )
+
+        return {}
+
+    def report_error(self, message):
+        self.request.session.flash(message, "error")
+        self.request.response.status_code = 400
```

Comment:
There wasn't a clear preferred pattern for making admin pages return 4xx pages in the event of a validation error, and IIRC some existing pages didn't handle this correctly in my testing.

Reply:
This should be handled using [exception views](https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/views.html#custom-exception-views) but I don't know if h's admin pages have those setup properly yet.

Reply:
Some pages make use of exception views, but it looks like this hasn't been set up in a standardized way yet. I think that would probably best be left as a separate refactoring exercise.

Reply:
:+1: